### PR TITLE
[SDTEST-3945] Optimize bulk file serialization

### PR DESCRIPTION
**Pull request by anmarchenko**
## Why

Serializing a large file list as separate Ruby normalization, deduplication, and MessagePack passes creates substantial transient arrays, hashes, and method calls. TIA coverage events are the current high-volume caller, but the performance concern is bulk file serialization rather than coverage collection.

This change introduces a dedicated `Datadog::CI::FileSerialization` Ruby module backed by its own `file_serialization.c` extension unit. After a cheap absolute-path pre-scan, its native fast path fuses repository-root validation and slicing, stable deduplication, and MessagePack encoding into one output pass. `DDCov` remains responsible only for coverage collection.

The current caller captures the repository root once because the root is a process invariant between coverage events. Unsupported packer or input shapes fall back to the existing Ruby implementation before any bytes are written.

## Benchmark

`ruby-rails-tia-stress-test-level`, four paired measurements with normal test tracing and no code coverage:

| | `main` | This PR |
|---|---:|---:|
| Baseline median | 37,014 ms | 36,149 ms |
| Instrumented median | 58,148 ms | 54,380 ms |
| Tracer overhead | 57.10% | 50.43% |

The overhead improves by **6.66 percentage points**. The paired runtime-multiplier comparison is conclusive: **95% CI −9.17% to −3.05%**.

## Correctness

- Native output is byte-for-byte compared with the legacy serializer for relative, absolute, duplicate, binary, Unicode, and long paths.
- 21 focused coverage-event examples and 47 DDCov examples passed (68 total).
- All 19,080 Rails benchmark assertions passed.
- Ruby 3.4.7 and Ruby 4.0.1 native builds passed.
- The extracted native extension compiles independently as part of the complete extension build.
- RBS missing/stale checks pass, Steep reports no type errors, and Standard reports no offenses across 506 files.

Correctness remains the priority: this does not skip coverage events or file classes, change ordering, or introduce nondeterministic behavior.

## Stack

This is PR 1 of 2 and targets `main`. #580 replaces the high-frequency NEWOBJ TracePoint wrapper with an owned raw event hook.

### CODE REVIEW

**Comment by anmarchenko** [comment]
@codex review

**Comment by anmarchenko** [comment]
@autotest review

**Comment by anmarchenko** [comment]
@codex review

**Comment by anmarchenko** [comment]
@codex review

File: lib/datadog/ci/test_impact_analysis/coverage/files.rb [hunk 1]
```diff
@@ -27,6 +32,30 @@ def size
...
+          # stable deduplication, and filename-entry packing. Any shape it does
+          # not support falls back before writing bytes.
+          def write_to(packer)
+            if FileSerialization.respond_to?(:pack_files) &&
+                (packed_files = FileSerialization.pack_files(
+                  @coverage,
+                  @custom_impacted_files,
+                  @root
+                ))
+              packer.buffer.write(packed_files)
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply lib/datadog/ci/test_impact_analysis/coverage/files.rb:46]: I specifically removed it because we only supply MessagePack ourselves and I know that we don't use compatibility mode

**Comment by anmarchenko** [comment]
## E2E Test Report: SUCCESS ✅

**Tested by:** Shepherd Agent (autonomous QA for Datadog Test Optimization)

### Test Environment

- **Method:** Local testing with the `rails-tia-stress` Ruby/RSpec playground and mockdog
- **Runtime:** Ruby 3.4.7 on macOS arm64
- **Revision:** `94b84de099510ad0e3fbfe5c17281d9a4d24f445`
- **Backend:** mockdog only, as requested

### Results

| Check | Status | Evidence |
| --- | --- | --- |
| Test-level TIA coverage | ✅ | 1,000 tests passed; 1,000 coverage records containing 2,010,000 file entries; all 19,080 assertions passed |
| Suite-level TIA coverage | ✅ | 1,000 tests passed; 100 coverage records containing 1,291,988 file entries; all 2,287 assertions passed |
| Coverage intake | ✅ | All 55 `/api/v2/citestcov` requests returned HTTP 200 (32 test-level + 23 suite-level) |
| Raw payload structure | ✅ | Gzipped multipart MessagePack v2; expected envelope/coverage/file keys; correct test/suite/session IDs |
| Coverage-to-span linkage | ✅ | Every raw test-level coverage record matched its exact raw test span; every suite-level record matched its exact suite span |
| File normalization | ✅ | Repository-relative, unique, non-empty paths; no traversal, absolute paths, backslashes, or NULs |
| Native byte parity | ✅ | Native output matched the legacy Ruby MessagePack serializer for real 2,010-file and 12,908-file payloads |
| Encoding and size boundaries | ✅ | Unicode, binary, long paths, stable deduplication, fallback cases, string lengths through 65,536 bytes, and array sizes through 65,536 entries |
| Randomized parity | ✅ | 500 deterministic randomized native-versus-legacy cases passed |
| Telemetry | ✅ | Valid telemetry with zero coverage errors and zero dropped payloads in both runs |

### Issues Found

No issues were found in PR #579.

Test-infrastructure note: mockdog's convenience `parsed/*.jsonl` representation rounds some 64-bit IDs after a JSON `float64` conversion. The raw `citestcycle.jsonl` and `citestcov.jsonl` records preserve exact IDs and matched correctly, so all linkage checks used the raw streams. This does not affect the PR's serialized coverage payloads.

### Verification

- Coverage payloads were decoded and recorded correctly by mockdog.
- Raw request bodies, event linkage, file contents, transport responses, and telemetry were inspected directly.
- Datadog UI verification was not performed because testing was explicitly limited to mockdog.

### Test Methodology

1. Injected the exact PR revision with Crook and enabled debug logging.
2. Ran both test-level and suite-level high-cardinality Rails/RSpec coverage profiles.
3. Executed the committed mockdog assertions.
4. Independently streamed and audited all raw coverage and test-cycle records.
5. Loaded the compiled PR native extension and compared its bytes with the legacy MessagePack serializer using real payloads, boundary cases, encoding cases, fallback cases, and a randomized corpus.

---

_This E2E test was performed by Shepherd - autonomous QA agent for Datadog Test Optimization_

**Review by calvinbayer — APPROVED** [review]
